File: fetch.go
package main

import (
	"crypto/tls"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"net/url"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

var (
	cachedData []StockData
	cacheMutex sync.RWMutex
	lastFetch  int64
	priceHist  = make(map[string][]float64)
	histMutex  sync.Mutex

	httpClient = &http.Client{
		Timeout: 8 * time.Second,
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
		},
	}
)

func initClients() {
	if appConfig.Proxy != "" {
		u, err := url.Parse(appConfig.Proxy)
		if err == nil {
			httpClient.Transport = &http.Transport{
				TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
				Proxy:           http.ProxyURL(u),
			}
		}
	}
	if appConfig.CacheTTL != 0 {
		cacheTTL = appConfig.CacheTTL
	}
}

var cacheTTL int64 = 55
var qtRe = regexp.MustCompile(`v_(\w+)="([^"]*)"`)

func neededStocks() []StockConfig {
	all := loadStocks()
	mode := GetViewMode()
	eff, isAuto := GetEffectiveGroup(mode)
	if eff == "ALL" {
		return all
	}
	if !isAuto {
		if eff == "" {
			return nil
		}
		var out []StockConfig
		for _, c := range all {
			if c.Group == eff {
				out = append(out, c)
			}
		}
		return out
	}
	matching := GetMatchingAutoGroups()
	if len(matching) == 0 {
		return nil
	}
	want := make(map[string]bool, len(matching))
	for _, g := range matching {
		want[g] = true
	}
	var out []StockConfig
	for _, c := range all {
		if want[c.Group] {
			out = append(out, c)
		}
	}
	return out
}

func fetchQT(codes []string) map[string][]string {
	out := make(map[string][]string)
	if len(codes) == 0 {
		return out
	}
	urlStr := "https://qt.gtimg.cn/q=" + strings.Join(codes, ",")
	req, _ := http.NewRequest("GET", urlStr, nil)
	req.Header.Set("Referer", "https://gu.qq.com/")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := httpClient.Do(req)
	if err != nil {
		return out
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	for _, m := range qtRe.FindAllSubmatch(body, -1) {
		code := string(m[1])
		raw := string(m[2])
		var vals []string
		if strings.Contains(raw, "~") {
			vals = strings.Split(raw, "~")
		} else {
			vals = strings.Split(raw, ",")
		}
		out[code] = vals
	}
	return out
}

func fetchYahoo(code string) (prices []float64, timestamps []int64, regStart, regEnd int64, price, prev, chartPrev float64) {
	urlStr := fmt.Sprintf("https://query1.finance.yahoo.com/v8/finance/chart/%s?interval=1m&range=1d", url.QueryEscape(code))
	req, _ := http.NewRequest("GET", urlStr, nil)
	req.Header.Set("User-Agent", "Mozilla/5.0")
	resp, err := httpClient.Do(req)
	if err != nil {
		return
	}
	defer resp.Body.Close()
	body, _ := io.ReadAll(resp.Body)
	var parsed struct {
		Chart struct {
			Result []struct {
				Meta struct {
					RegularMarketPrice float64 `json:"regularMarketPrice"`
					PreviousClose      float64 `json:"previousClose"`
					ChartPreviousClose float64 `json:"chartPreviousClose"`
					CurrentTradingPeriod struct {
						Regular struct {
							Start int64 `json:"start"`
							End   int64 `json:"end"`
						} `json:"regular"`
					} `json:"currentTradingPeriod"`
				} `json:"meta"`
				Timestamp  []int64 `json:"timestamp"`
				Indicators struct {
					Quote []struct {
						Close []*float64 `json:"close"`
					} `json:"quote"`
				} `json:"indicators"`
			} `json:"result"`
		} `json:"chart"`
	}
	if err := json.Unmarshal(body, &parsed); err != nil || len(parsed.Chart.Result) == 0 {
		return
	}
	res := parsed.Chart.Result[0]
	if len(res.Indicators.Quote) > 0 && len(res.Timestamp) > 0 {
		closes := res.Indicators.Quote[0].Close
		for i, v := range closes {
			if i >= len(res.Timestamp) {
				break
			}
			if v != nil {
				prices = append(prices, *v)
				timestamps = append(timestamps, res.Timestamp[i])
			}
		}
	} else if len(res.Indicators.Quote) > 0 {
		for _, v := range res.Indicators.Quote[0].Close {
			if v != nil {
				prices = append(prices, *v)
			}
		}
	}
	price = res.Meta.RegularMarketPrice
	if price == 0 && len(prices) > 0 {
		price = prices[len(prices)-1]
	}
	prev = res.Meta.PreviousClose
	if prev == 0 {
		prev = res.Meta.ChartPreviousClose
	}
	chartPrev = res.Meta.ChartPreviousClose
	regStart = res.Meta.CurrentTradingPeriod.Regular.Start
	regEnd = res.Meta.CurrentTradingPeriod.Regular.End
	return
}

func parseGtimgRows(rows []string) []float64 {
	prices := make([]float64, 0, len(rows))
	for _, row := range rows {
		parts := strings.Fields(row)
		if len(parts) < 2 {
			continue
		}
		p, err := strconv.ParseFloat(parts[1], 64)
		if err != nil {
			continue
		}
		prices = append(prices, p)
	}
	if len(prices) < 2 {
		return nil
	}
	return prices
}

func fetchGtimgMinute(code string) []float64 {
	isUS := strings.HasPrefix(code, "us")
	var urlStr string
	if isUS {
		urlStr = "https://web.ifzq.gtimg.cn/appstock/app/UsMinute/query?code=" + url.QueryEscape(code)
	} else {
		urlStr = "https://web.ifzq.gtimg.cn/appstock/app/minute/query?code=" + code
	}
	req, _ := http.NewRequest("GET", urlStr, nil)
	req.Header.Set("User-Agent", "Mozilla/5.0")
	if isUS {
		req.Header.Set("Referer", "https://gu.qq.com/")
	}
	resp, err := httpClient.Do(req)
	if err != nil {
		return nil
	}
	defer resp.Body.Close()
	var j map[string]interface{}
	if err := json.NewDecoder(resp.Body).Decode(&j); err != nil {
		return nil
	}
	if isUS {
		if codeVal, ok := j["code"].(float64); ok && codeVal != 0 {
			return nil
		}
	}
	dataMap, _ := j["data"].(map[string]interface{})
	item, _ := dataMap[code].(map[string]interface{})
	d, _ := item["data"].(map[string]interface{})
	rowsIf, _ := d["data"].([]interface{})
	rows := make([]string, 0, len(rowsIf))
	for _, r := range rowsIf {
		if s, ok := r.(string); ok {
			rows = append(rows, s)
		}
	}
	prices := parseGtimgRows(rows)
	if isUS && prices != nil {
		filtered := prices[:0]
		for _, p := range prices {
			if p > 0 {
				filtered = append(filtered, p)
			}
		}
		if len(filtered) < 2 {
			return nil
		}
		prices = filtered
	}
	return prices
}

func parseQT(code string, vals []string) (price, change, pct, prev float64) {
	defer func() {
		if r := recover(); r != nil {
			price, change, pct, prev = 0, 0, 0, 0
		}
	}()
	if strings.HasPrefix(code, "hf_") {
		if len(vals) > 0 {
			price, _ = strconv.ParseFloat(vals[0], 64)
		}
		if len(vals) > 1 {
			change, _ = strconv.ParseFloat(vals[1], 64)
		}
		prev = price - change
		if prev != 0 {
			pct = (change / prev) * 100
		}
	} else {
		if len(vals) > 3 {
			price, _ = strconv.ParseFloat(vals[3], 64)
		}
		if len(vals) > 4 {
			prev, _ = strconv.ParseFloat(vals[4], 64)
		}
		if prev == 0 {
			prev = price
		}
		if len(vals) > 32 {
			pct, _ = strconv.ParseFloat(vals[32], 64)
		}
		if len(vals) > 31 {
			change, _ = strconv.ParseFloat(vals[31], 64)
		} else {
			change = price - prev
		}
	}
	return
}

type chartResult struct {
	prices     []float64
	timestamps []int64
	regStart   int64
	regEnd     int64
	chartPrev  float64
	yPrice     float64
	yPrev      float64
}

func refreshData() []StockData {
	configs := neededStocks()
	if len(configs) != len(loadStocks()) {
		eff, _ := GetEffectiveGroup(GetViewMode())
		log.Printf("[fetch] View %s: fetching %d/%d stocks", eff, len(configs), len(loadStocks()))
	}
	var tencentCodes []string
	for _, c := range configs {
		if c.Source == "tencent" {
			tencentCodes = append(tencentCodes, c.Code)
		}
	}
	qt := fetchQT(tencentCodes)

	chartMap := make(map[string]chartResult)
	var wg sync.WaitGroup
	var mu sync.Mutex
	for _, c := range configs {
		wg.Add(1)
		go func(cfg StockConfig) {
			defer wg.Done()
			var res chartResult
			switch cfg.Source {
			case "tencent":
				res.prices = fetchGtimgMinute(cfg.Code)
			case "yahoo":
				res.prices, res.timestamps, res.regStart, res.regEnd, res.yPrice, res.yPrev, res.chartPrev = fetchYahoo(cfg.Code)
			}
			mu.Lock()
			chartMap[cfg.Code] = res
			mu.Unlock()
		}(c)
	}
	wg.Wait()

	var items []StockData
	for _, c := range configs {
		code := c.Code
		cRes := chartMap[code]
		var price, change, pct, prev float64
		var prices []float64

		switch c.Source {
		case "tencent":
			vals := qt[code]
			if len(vals) == 0 {
				clean := strings.ReplaceAll(strings.ReplaceAll(code, "sh", ""), "sz", "")
				vals = qt[clean]
			}
			price, change, pct, prev = parseQT(code, vals)
			if code == "^VIX" {
				price, change, pct, prev = 0, 0, 0, 0
			}
			prices = cRes.prices
		case "yahoo":
			price, prev = cRes.yPrice, cRes.yPrev
			if prev != 0 {
				change = price - prev
				pct = (change / prev) * 100
			}
			prices = cRes.prices
			if code == "^VIX" && price == 0 {
				price, change, pct, prev = 0, 0, 0, 0
			}
		}

		isChart := false
		if c.Source == "tencent" {
			isChart = strings.HasPrefix(code, "sh") || strings.HasPrefix(code, "sz") || strings.HasPrefix(code, "us")
		} else {
			isChart = len(prices) >= 2 || price > 0
		}
		if !isChart {
			prices = nil
		} else {
			histMutex.Lock()
			h := priceHist[code]
			if price > 0 {
				h = append(h, price)
				if len(h) > 1440 {
					h = h[len(h)-1440:]
				}
				priceHist[code] = h
			}
			if len(prices) < 2 && len(h) > 2 {
				prices = append([]float64(nil), h...)
			}
			histMutex.Unlock()
		}

		items = append(items, StockData{
			Code: code, Name: c.Name, Group: c.Group,
			Price: price, Change: change, Pct: pct, Prev: prev,
			Prices: prices,
			Timestamps: cRes.timestamps, RegularStart: cRes.regStart, RegularEnd: cRes.regEnd, ChartPrevClose: cRes.chartPrev,
		})
	}
	cacheMutex.Lock()
	cachedData = items
	lastFetch = time.Now().Unix()
	cacheMutex.Unlock()
	return items
}

func getData() []StockData {
	needed := neededStocks()
	if len(needed) == 0 {
		return nil
	}
	needSet := make(map[string]bool, len(needed))
	for _, c := range needed {
		needSet[c.Code] = true
	}
	cacheMutex.RLock()
	now := time.Now().Unix()
	ttl := cacheTTL
	if appConfig.CacheTTL != 0 {
		ttl = appConfig.CacheTTL
	}
	if now-lastFetch <= ttl && len(cachedData) > 0 {
		cachedCodes := make(map[string]bool, len(cachedData))
		for _, d := range cachedData {
			cachedCodes[d.Code] = true
		}
		hit := true
		for code := range needSet {
			if !cachedCodes[code] {
				hit = false
				break
			}
		}
		if hit {
			defer cacheMutex.RUnlock()
			return cachedData
		}
	}
	cacheMutex.RUnlock()
	return refreshData()
}

File: kkanpan-rs/src/fetch.rs
use crate::config::{self, StockConfig, StockData};
use std::collections::HashMap;
use std::sync::{Mutex, RwLock};

// ---- 缓存（对齐 fetch.go） ----

static CACHED: RwLock<Vec<StockData>> = RwLock::new(Vec::new());
static LAST_FETCH: RwLock<i64> = RwLock::new(0);
static PRICE_HIST: std::sync::LazyLock<Mutex<HashMap<String, Vec<f64>>>> =
    std::sync::LazyLock::new(|| Mutex::new(HashMap::new()));

fn now_unix() -> i64 {
    std::time::SystemTime::now()
        .duration_since(std::time::UNIX_EPOCH)
        .map(|d| d.as_secs() as i64)
        .unwrap_or(0)
}

// ---- HTTP：默认 curl 子进程（Kindle 自带 curl，零 TLS 编译风险） ----

fn curl_get(url: &str, referer: &str) -> Option<Vec<u8>> {
    let mut cmd = std::process::Command::new("curl");
    cmd.args(["-sk", "--max-time", "8", "-A", "Mozilla/5.0"]);
    let proxy = &config::app().proxy;
    if !proxy.is_empty() {
        cmd.args(["-x", proxy]);
    }
    if !referer.is_empty() {
        cmd.arg("-H").arg(format!("Referer: {referer}"));
    }
    cmd.arg(url);
    let out = cmd.output().ok()?;
    if out.status.success() { Some(out.stdout) } else { None }
}

#[cfg(feature = "net-ureq")]
fn ureq_get(url: &str, referer: &str) -> Option<Vec<u8>> {
    // 第二步用：切到 ureq+ring（仅 x86 开发机 / 验证机；armv7 上 ring 可编但慢）。
    // 注意与 Go 版差异：此处默认校验证书（Go 版 InsecureSkipVerify）。
    let mut b =
        ureq::Agent::config_builder().timeout_global(Some(std::time::Duration::from_secs(8)));
    let proxy = &config::app().proxy;
    if !proxy.is_empty() {
        b = b.proxy(ureq::Proxy::new(proxy).ok()?);
    }
    let agent = b.build().new_agent();
    let mut resp = agent
        .get(url)
        .header("User-Agent", "Mozilla/5.0")
        .header("Referer", referer)
        .call()
        .ok()?;
    let mut buf = Vec::new();
    std::io::copy(resp.body_mut().as_reader(), &mut buf).ok()?;
    Some(buf)
}

#[cfg(not(feature = "net-ureq"))]
fn http_get(url: &str, referer: &str) -> Option<Vec<u8>> {
    curl_get(url, referer)
}

#[cfg(feature = "net-ureq")]
fn http_get(url: &str, referer: &str) -> Option<Vec<u8>> {
    ureq_get(url, referer)
}

// ---- 抓取 ----

fn needed_stocks(view: &str) -> Vec<StockConfig> {
    let all = config::stocks();
    let (eff, is_auto) = config::effective_group(view);
    if eff == "ALL" {
        return all.to_vec();
    }
    if !is_auto {
        if eff.is_empty() {
            return vec![];
        }
        return all.iter().filter(|c| c.group == eff).cloned().collect();
    }
    let matching = config::matching_auto_groups();
    if matching.is_empty() {
        return vec![];
    }
    all.iter()
        .filter(|c| matching.contains(&c.group))
        .cloned()
        .collect()
}

// 腾讯批量行情：v_code="..."，~ 或 , 分隔
fn fetch_qt(codes: &[String]) -> HashMap<String, Vec<String>> {
    let mut out = HashMap::new();
    if codes.is_empty() {
        return out;
    }
    let url = format!("https://qt.gtimg.cn/q={}", codes.join(","));
    let body = match http_get(&url, "https://gu.qq.com/") {
        Some(b) => String::from_utf8_lossy(&b).into_owned(),
        None => return out,
    };
    for seg in body.split("v_").skip(1) {
        let Some(eq) = seg.find("=\"") else { continue };
        let code = &seg[..eq];
        let rest = &seg[eq + 2..];
        let Some(end) = rest.find('"') else { continue };
        let raw = &rest[..end];
        let vals: Vec<String> = if raw.contains('~') {
            raw.split('~').map(|s| s.to_string()).collect()
        } else {
            raw.split(',').map(|s| s.to_string()).collect()
        };
        out.insert(code.to_string(), vals);
    }
    out
}

struct ChartResult {
    prices: Vec<f64>,
    timestamps: Vec<i64>,
    reg_start: i64,
    reg_end: i64,
    chart_prev: f64,
    y_price: f64,
    y_prev: f64,
}

fn fetch_yahoo(code: &str) -> ChartResult {
    let mut r = ChartResult {
        prices: vec![],
        timestamps: vec![],
        reg_start: 0,
        reg_end: 0,
        chart_prev: 0.0,
        y_price: 0.0,
        y_prev: 0.0,
    };
    let url = format!(
        "https://query1.finance.yahoo.com/v8/finance/chart/{}?interval=1m&range=1d",
        code
    );
    let body = match http_get(&url, "") {
        Some(b) => b,
        None => return r,
    };
    let v: serde_json::Value = match serde_json::from_slice(&body) {
        Ok(v) => v,
        Err(_) => return r,
    };
    let res = match v.pointer("/chart/result/0") {
        Some(v) => v,
        None => return r,
    };
    r.y_price = res.pointer("/meta/regularMarketPrice").and_then(|x| x.as_f64()).unwrap_or(0.0);
    r.y_prev = res.pointer("/meta/previousClose").and_then(|x| x.as_f64()).unwrap_or(0.0);
    r.chart_prev = res.pointer("/meta/chartPreviousClose").and_then(|x| x.as_f64()).unwrap_or(0.0);
    if r.y_prev == 0.0 {
        r.y_prev = r.chart_prev;
    }
    r.reg_start = res.pointer("/meta/currentTradingPeriod/regular/start").and_then(|x| x.as_i64()).unwrap_or(0);
    r.reg_end = res.pointer("/meta/currentTradingPeriod/regular/end").and_then(|x| x.as_i64()).unwrap_or(0);
    let empty = vec![];
    let stamps = res.pointer("/timestamp").and_then(|x| x.as_array()).unwrap_or(&empty);
    let closes = res.pointer("/indicators/quote/0/close").and_then(|x| x.as_array());
    if let Some(cl) = closes {
        for (i, c) in cl.iter().enumerate() {
            if let Some(p) = c.as_f64() {
                r.prices.push(p);
                if let Some(t) = stamps.get(i).and_then(|x| x.as_i64()) {
                    r.timestamps.push(t);
                }
            }
        }
    }
    if r.y_price == 0.0 {
        if let Some(&last) = r.prices.last() {
            r.y_price = last;
        }
    }
    r
}

// 腾讯分时："HH:MM 价格 成交量"
fn parse_gtimg_rows(rows: &[&str]) -> Option<Vec<f64>> {
    let mut prices = Vec::with_capacity(rows.len());
    for row in rows {
        let mut it = row.split_whitespace();
        it.next()?;
        let p: f64 = it.next()?.parse().ok()?;
        prices.push(p);
    }
    if prices.len() < 2 { None } else { Some(prices) }
}

fn fetch_gtimg_minute(code: &str) -> Option<Vec<f64>> {
    let is_us = code.starts_with("us");
    let url = if is_us {
        format!("https://web.ifzq.gtimg.cn/appstock/app/UsMinute/query?code={code}")
    } else {
        format!("https://web.ifzq.gtimg.cn/appstock/app/minute/query?code={code}")
    };
    let body = http_get(&url, if is_us { "https://gu.qq.com/" } else { "" })?;
    let v: serde_json::Value = serde_json::from_slice(&body).ok()?;
    if is_us && v.pointer("/code").and_then(|x| x.as_f64()).unwrap_or(0.0) != 0.0 {
        return None;
    }
    let rows = v.pointer(&format!("/data/{code}/data/data"))?.as_array()?;
    let strs: Vec<&str> = rows.iter().filter_map(|x| x.as_str()).collect();
    let mut prices = parse_gtimg_rows(&strs)?;
    if is_us {
        prices.retain(|&p| p > 0.0);
        if prices.len() < 2 {
            return None;
        }
    }
    Some(prices)
}

fn parse_qt(code: &str, vals: &[String]) -> (f64, f64, f64, f64) {
    if code == "^VIX" {
        return (0.0, 0.0, 0.0, 0.0);
    }
    let f = |i: usize| vals.get(i).and_then(|s| s.parse::<f64>().ok()).unwrap_or(0.0);
    if code.starts_with("hf_") {
        let (price, change) = (f(0), f(1));
        let prev = price - change;
        let pct = if prev != 0.0 { change / prev * 100.0 } else { 0.0 };
        return (price, change, pct, prev);
    }
    let price = f(3);
    let mut prev = f(4);
    if prev == 0.0 {
        prev = price;
    }
    let pct = if vals.len() > 32 { f(32) } else { 0.0 };
    let change = if vals.len() > 31 { f(31) } else { price - prev };
    (price, change, pct, prev)
}

pub fn refresh_data(view: &str) -> Vec<StockData> {
    let configs = needed_stocks(view);
    let tencent: Vec<String> = configs
        .iter()
        .filter(|c| c.source == "tencent")
        .map(|c| c.code.clone())
        .collect();
    let qt = fetch_qt(&tencent);

    // 分时并行（std::thread::scope，无额外依赖）
    let charts: HashMap<String, ChartResult> = std::thread::scope(|s| {
        let handles: Vec<_> = configs
            .iter()
            .map(|c| {
                s.spawn(|| {
                    let mut r = ChartResult {
                        prices: vec![],
                        timestamps: vec![],
                        reg_start: 0,
                        reg_end: 0,
                        chart_prev: 0.0,
                        y_price: 0.0,
                        y_prev: 0.0,
                    };
                    if c.source == "tencent" {
                        r.prices = fetch_gtimg_minute(&c.code).unwrap_or_default();
                    } else if c.source == "yahoo" {
                        r = fetch_yahoo(&c.code);
                    }
                    (c.code.clone(), r)
                })
            })
            .collect();
        handles.into_iter().map(|h| h.join().unwrap()).collect()
    });

    let mut items = Vec::with_capacity(configs.len());
    for c in &configs {
        let cr = charts.get(&c.code);
        let (price, change, pct, prev, mut prices) = if c.source == "tencent" {
            let mut vals = qt.get(&c.code).cloned().unwrap_or_default();
            if vals.is_empty() {
                let clean = c.code.replace("sh", "").replace("sz", "");
                vals = qt.get(&clean).cloned().unwrap_or_default();
            }
            let (p, ch, pc, pr) = parse_qt(&c.code, &vals);
            (p, ch, pc, pr, cr.map(|r| r.prices.clone()).unwrap_or_default())
        } else {
            let (yp, ypr) = cr.map(|r| (r.y_price, r.y_prev)).unwrap_or((0.0, 0.0));
            let (ch, pc) = if ypr != 0.0 {
                (yp - ypr, (yp - ypr) / ypr * 100.0)
            } else {
                (0.0, 0.0)
            };
            let (p, ch, pc, pr) = if c.code == "^VIX" && yp == 0.0 {
                (0.0, 0.0, 0.0, 0.0)
            } else {
                (yp, ch, pc, ypr)
            };
            (p, ch, pc, pr, cr.map(|r| r.prices.clone()).unwrap_or_default())
        };

        let is_chart = if c.source == "tencent" {
            c.code.starts_with("sh") || c.code.starts_with("sz") || c.code.starts_with("us")
        } else {
            prices.len() >= 2 || price > 0.0
        };
        if !is_chart {
            prices.clear();
        } else {
            let mut hist = PRICE_HIST.lock().unwrap();
            let h = hist.entry(c.code.clone()).or_default();
            if price > 0.0 {
                h.push(price);
                if h.len() > 1440 {
                    let n = h.len() - 1440;
                    h.drain(..n);
                }
            }
            if prices.len() < 2 && h.len() > 2 {
                prices = h.clone();
            }
        }

        items.push(StockData {
            code: c.code.clone(),
            name: c.name.clone(),
            group: c.group.clone(),
            price,
            change,
            pct,
            prev,
            prices,
            timestamps: cr.map(|r| r.timestamps.clone()).unwrap_or_default(),
            regular_start: cr.map(|r| r.reg_start).unwrap_or(0),
            regular_end: cr.map(|r| r.reg_end).unwrap_or(0),
            chart_prev_close: cr.map(|r| r.chart_prev).unwrap_or(0.0),
        });
    }
    *CACHED.write().unwrap() = items.clone();
    *LAST_FETCH.write().unwrap() = now_unix();
    items
}

pub fn get_data(view: &str) -> Vec<StockData> {
    let needed = needed_stocks(view);
    if needed.is_empty() {
        return vec![];
    }
    let ttl = config::app().cache_ttl;
    {
        let cached = CACHED.read().unwrap();
        if now_unix() - *LAST_FETCH.read().unwrap() <= ttl && !cached.is_empty() {
            let have: std::collections::HashSet<&str> =
                cached.iter().map(|d| d.code.as_str()).collect();
            if needed.iter().all(|c| have.contains(c.code.as_str())) {
                return cached.clone();
            }
        }
    }
    refresh_data(view)
}

pub fn cached_snapshot() -> Vec<StockData> {
    CACHED.read().unwrap().clone()
}
